smartsim/_core/_cli/build.py

```diff
@@ -331,84 +328,146 @@ def infer_torch_device(self) -> str:
-        try:
-            if not self.build_env.check_installed("onnx", self.versions.ONNX):
-                msg = (
-                    f"ONNX {self.versions.ONNX} not installed in python environment. "
-                    + f"Consider installing onnx=={self.versions.ONNX} with pip"
-                )
-                logger.warning(msg)
-            else:
-                logger.info(
-                    f"ONNX {self.versions.ONNX} installed in Python environment"
-                )
-        except SetupError as e:
-            logger.warning(str(e))
+        self._check_py_package_version(
+            {
+                "onnx": f"{self.versions.ONNX}",
+                "skl2onnx": f"{self.versions.REDISAI.skl2onnx}",
+                "onnxmltools": f"{self.versions.REDISAI.onnxmltools}",
+                "scikit-learn": f"{self.versions.REDISAI.__getattr__('scikit-learn')}",
+            },
+            install_on_absent=handle_conflict,
+            install_on_conflict=handle_conflict,
+        )
 
-    def check_tf_install(self) -> None:
-        """Check Python environment for TensorFlow installation"""
+    def check_py_tf_version(self, handle_conflict: bool = False) -> None:
+        """Check Python environment for a compatible TensorFlow installation"""
+        logger.info(f"Searching for a compatible TF install...")
+        self._check_py_package_version(
+            {"tensorflow": self.versions.TENSORFLOW},
+            install_on_absent=handle_conflict,
+            install_on_conflict=handle_conflict,
+        )
 
-        try:
-            if not self.build_env.check_installed(
-                "tensorflow", self.versions.TENSORFLOW
-            ):
-                msg = (
-                    f"TensorFlow {self.versions.TENSORFLOW} not installed in Python environment. "
-                    + f"Consider installing tensorflow=={self.versions.TENSORFLOW} with pip"
-                )
-                logger.warning(msg)
-            else:
-                logger.info(
-                    f"TensorFlow {self.versions.TENSORFLOW} installed in Python environment"
-                )
-        except SetupError as e:
-            logger.warning(str(e))
+    def _check_py_package_version(
+        self,
+        packages: t.Mapping[str, t.Optional[str]],
+        end_point: t.Optional[str] = None,
+        validator: t.Optional[t.Callable[[str, t.Optional[str]], bool]] = None,
+        install_on_absent: bool = False,
+        install_on_conflict: bool = False,
+    ) -> None:
+        missing: t.List[str] = []
+        conflicts: t.List[str] = []
+        to_uninstall: t.List[str] = []
+        to_install: t.List[str] = []
+        validator = validator or self.build_env.check_installed
+        verbose_info: t.Callable[[str], t.Any] = lambda s: ...
+        if self.verbose:
+            verbose_info = logger.info
+
+        for name, version in packages.items():
+            spec = f"{name}=={version}" if version else name
```

Comment:
How are we going to handle requests for package versions 3.4.x or for version >= 2.6.0?

Reply:
Good point! Now that we know that the RAI versions are more suggestions than hard pins, we could probably work on allowing more flexible versions. I did some preliminary work to get the ball rolling on this, but more thought could probably go into this! Adding a ticket!